std/experimental/allocator/package.d

```diff
@@ -226,6 +226,18 @@ module std.experimental.allocator;
 public import std.experimental.allocator.common,
     std.experimental.allocator.typed;
 
+// Fix issue 17806: this should always be the first unittest in this module
+// in order to ensure that we use the `processAllocator` setter before the getter
+@system unittest
+{
+    import std.experimental.allocator.mallocator: Mallocator;
+    import std.experimental.allocator.gc_allocator : GCAllocator;
+    auto newAlloc = sharedAllocatorObject(Mallocator.instance);
+    processAllocator = newAlloc;
+    assert(processAllocator is newAlloc);
+    processAllocator = sharedAllocatorObject(GCAllocator.instance);
```

Comment:
Maybe not implicitly assume that it was the GCAllocator before and just reset it to the old value?

Reply:
This would require the setter to return the previous allocator, which it currently doesn't. In order to change the API to do this, we'd need to guard with a `lock`. We can regard this as a best effort.

Reply:
>This would require the setter to return the previous allocator, which it currently doesn't.

That would also be a potential landmine since for basic types the result of `(a = b)` is the post-modification value of `a`, not the pre-modification value of `a`.